### ISSUE TITLE
[Feature request] Auto-optimize ld to ldh whenever possible

### ISSUE THREAD

**Issue by ISSOtm** [rednex/rgbds#229]
I know this has been brought up multiple times. This issue is open to discuss a possible implementation, not request the feature.

My idea is to add a pass before section locations are determined, but after object files have been read.
This pass would look at all `ld` operations with a label (which should be doable by parsing the label patch list, and adding a flag telling if the patch applies to an operation that could be optimized. Said flag could be re-used for other optimizations, such as `jp`->`jr`, **though this one is out of this issue's scope**).
Each time, the label is looked up (either by constructing a table of HRAM labels while reading object files, or by searching all object files for the label each time), and if it's in HRAM, then the instruction is modified in the section data into the corresponding `ldh` instruction, and the section size would be updated accordingly.
Of course, there would be a flag to turn off this optimization ; it could even be done at assembling time, since the assembler could simply omit the flag marking the instruction for optimization.

Perks :
- This doesn't change the flow of the linking
- Data is passed seamlessly to the next linking passes (since this would edit the object data in-place)
- This isn't an ugly hack (although some may complain about the "marked for optimization" bit)
- This would be a convenience, for example solving [pokered's HRAM file](https://github.com/pret/pokered/blob/master/hram.asm) containing only constants instead of labels (which means they are absent from the SYM file)
- Possibility to toggle it on a per-file basis (there could be use cases), or maybe on a per-instruction basis, with an extra directive

Downsides :
- The object format isn't really planned to be re-written after assembling, so this may lead to copying blocks of data around.
- There already is a "fix", which is using `ldh` everywhere, but it's easy to forget one (especially when moving variables from WRAM to HRAM during development)

I will attempt to implement this, and opened this issue as a request for ideas, comments, and critics.

**Comment by aaaaaa123456789** [comment]
Strong -1 here. There are ways of solving this issue in sane ways (I proposed one in #223) that aren't as harmful as a patch that allows resizing sections.

Consider code such as this:

```
  ld a, [hFoo]
  and a
  jr nz, .skip
  ld a, 1
  ld [hFoo], a
.skip```

This seems harmless, but resizing the section makes that `jr` jump to the wrong place — resizing the section also requires shifting the code around, since you'd be deleting a byte in the middle. Thus, every single usage of every single label would have to be tracked and solved by the linker. Which in turn brings its own set of problems, since expressions that aren't constant at the time of linking are limited in use.

Additionally, consider this other example:

```
  jr nc, .somewhere
  ld a, [hFoo]
  ; 125 bytes omitted
.somewhere```

If `hFoo` is only resolved at link time, this code would fail to compile, because the jump target would be too far.

I'm sure there are other issues I'm not thinking of right away. In the end, any solution to the HRAM problem must make the value of the symbol known _at assembly time_ somehow.

**Comment by ISSOtm** [comment]
The jr would still jump to the right place, since it's linked to the label. Furthermore, everything label-related is (should) be resolved at link time, thus the code would successfully compile (because the `ld` would first be resolved as a `ldh`), and only then would the `jr`'s offset be calculated and found correct.
If `jr` offsets are calculated at assembly time, and I suspect they do, then that work should be passed onto the linker.

The very concept of linking is that assembly must be independent from other files. Thus, making labels known at assembly time would completely defeat its purpose (and break `make`, as an extra bonus).

**Comment by aaaaaa123456789** [comment]
Considering the large amount of code that relies on the programmer knowing exactly the number of bytes that each line will generate (this is not too far-fetched when you consider the little space available in ROM and the possibility of copying code to RAM), I'd be very wary of letting the linker resize sections at will. This is a slippery slope.

If you're really considering implementing features like this, I'll kindly ask that you require a flag to enable them. This is a null feature (it adds no new behavior, as the behavior you ask for can easily be requested by using `ldh`) that brings in significant risks/drawbacks. This is the same reason I've been standing against an automatic `jp` to `jr` conversion (but I wouldn't oppose to some `jx` pseudo-op that does the selection) — behavior that breaks user expectations should be opt-in, not opt-out.

**Comment by ISSOtm** [comment]
In the first comment, I wrote
  [...] which should be doable by parsing the label patch list, and adding a flag telling if the patch applies to an operation that could be optimized.
I meant a command-line flag. I agree that enabling this behavior by default would break compatibility.

Letting the linker resize sections could be a slippery slope, except if all behaviors that imply section resizing ensure that they don't increase the size (such as replacing a 3-byte `ld` with a 2-byte `ldh`, etc.)

I'm doubtful of the automatic `jp`-to-`jr` because it depends on the memory layout, so it has first to link, check if an optimization is possible, and if so, re-link. I'm afraid of infinite loops that may occur, so I think it would be better to leave this one out entirely. `ld`-to-`ldh` is static, and linking can't fail because of it. That's why I'm leaving both of them in separate categories.

**Comment by AntonioND** [comment]
I think I'm going to reject this. Let's be honest, how many times do you really want to optimize a `ld` to `ldh`? That's only an issue for I/O and HRAM:

- I/O addresses are constant, so the assembler will know it has to use `ldh` (and the developer can use it explicitly for peace of mind).
- A developer only places a few things in HRAM, usually to optimize, and the normal thing to do is to keep track of all of them so that all `ld` are actually `ldh`.

> I'm doubtful of the automatic jp-to-jr because it depends on the memory layout, so it has first to link, check if an optimization is possible, and if so, re-link. I'm afraid of infinite loops that may occur, so I think it would be better to leave this one out entirely.

Yes, that's a good argument to be against this optimization too.

In short, I'm against both optimizations. Actually, the fact that rgbasm can choose between `ld` and `ldh` is a bit evil, and I was thinking about removing that choice and forcing the coder to use `ldh` if needed.

I've created this new issue: https://github.com/rednex/rgbds/issues/243
